### ISSUE TITLE
BatchNorm with mixed_bfloat16 returns empty output

### ISSUE THREAD

**Issue by cchan-lm** [tensorflow/tensorflow#55738]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.7.0

### Custom Code

No

### OS Platform and Distribution

Linx Ubuntu 18.04

### Python version

3.8.10

### CUDA/cuDNN version

11.2/8.1

### GPU model and memory

NVIDIA A100

### Current Behaviour?

```shell
BatchNorm layer using `dtype="mixed_bfloat16"` or "bfloat16" on an NVIDIA A100 returns an empty output with TF 2.7.0. However, when using `dtype="float32"`, "float16", or "mixed_float16", an actual input is returned.

As a side note, it is also only with `float32` that I get the message `tensorflow/stream_executor/cuda/cuda_dnn.cc:366] Loaded cuDNN version 8100`. `mixed_float16` gives the message `Your GPU will likely run quickly with dtype policy mixed_float16 as it has compute capability of at least 7.0. Your GPU: A100-PCIE-40GB, compute capability 8.0`. `float16` does not have a message of its own, it seems.

To minimize environmental problems, I used the `tensorflow/tensorflow:2.7.0-gpu` Docker image and ran `docker run -it --gpus all tensorflow/tensorflow:2.7.0-gpu bash`

Thank you in advance for helping solve this issue!
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

test_bn = tf.keras.layers.BatchNormalization(dtype="mixed_bfloat16")
test_bn(np.random.randn(16, 24, 24, 32))
```

### Relevant log output

```shell
When using mixed_bfloat16 or bfloat16:

2022-04-25 18:55:54.298268: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-04-25 18:55:54.845911: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38403 MB memory:  -> device: 0, name: A100-PCIE-40GB, pci bus id: 0000:01:00.0, compute capability: 8.0
<tf.Tensor: shape=(0,), dtype=float32, numpy=array([], dtype=float32)>


When using float32:

2022-04-25 18:58:51.935409: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-04-25 18:58:53.684462: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38403 MB memory:  -> device: 0, name: A100-PCIE-40GB, pci bus id: 0000:01:00.0, compute capability: 8.0
2022-04-25 18:58:54.296455: I tensorflow/stream_executor/cuda/cuda_dnn.cc:366] Loaded cuDNN version 8100
<tf.Tensor: shape=(16, 24, 24, 32), dtype=float32, numpy=
array([[[[-1.78168976e+00,  8.01364720e-01,  3.17569762e-01, ...,
          -1.08890069e+00, -4.05266196e-01,  9.75035071e-01],
         [-9.24855590e-01,  5.21459460e-01,  1.35295260e+00, ...,
          -1.89387918e+00,  8.45147550e-01, -7.09079564e-01],
         [ 1.49759853e+00,  3.18406433e-01,  2.32584342e-01, ...,
           2.76129186e-01, -2.41327822e-01, -4.68785256e-01],
         ...,

...
```
</details>

**Comment by sushreebarsa** [comment]
@cchan-lm I was able to replicate this issue on colab , please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/bd11bccbff7fa63c1e105f8274961701/55738.ipynb#scrollTo=tDOrSM8M7Nn7).Could you please refer [this ](https://stackoverflow.com/questions/58612783/batch-normalization-yes-or-no)thread .
For any further queries 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thanks!

**Comment by cchan-lm** [comment]
Thank you, @sushreebarsa! I noticed in the gist that using FP32 was not tested. Did you see that actual results were returned with FP32? The thread you linked seems to discuss on its numerical inconsistencies, not empty output. Was there something I missed? I tried a batch size of 32 to find the same results as well.

**Comment by sushreebarsa** [comment]
@cchan-lm Yes , I was also able to replicate the issue as reported here.Please find the updated [gist](https://colab.research.google.com/gist/sushreebarsa/bd11bccbff7fa63c1e105f8274961701/55738.ipynb#scrollTo=DAqTDD5stWF4) .Could you please confirm the same ?
Thanks!

**Comment by cchan-lm** [comment]
@sushreebarsa Thanks! Yep, that is what I found. Is this a TF problem or a Keras problem? It's very odd that it returns empty on one dtype but not others.

**Comment by reedwm** [comment]
Outside XLA, TF doesn't yet support bfloat16 on GPUs. However, an error should be raised, or the op should be run on the CPU if the CPU version supports bfloat16. It should not return an empty output, so I'll look into this.

I would recommend using float16 instead of bfloat16 on GPUs. 

Alternatively, you can use XLA by passing `jit_compile=True` to a `tf.function`, as XLA has some support for bfloat16 (but note you won't be able to switch to eager mode for debugging). See [the tutorial](https://www.tensorflow.org/xla/tutorials/jit_compile) for more info. E.g., the following works:

```python
test_bn = tf.keras.layers.BatchNormalization(dtype="mixed_bfloat16")
@tf.function(jit_compile=True)
def run_test_bn():
  return test_bn(np.random.randn(16, 24, 24, 32))
print(run_test_bn())
```

**Comment by cchan-lm** [comment]
@reedwm Thanks! That is very interesting! For my team's purposes, we are using bfloat16. I see the same successful result as you using XLA for BatchNorm. However, why is it not needed for Dense on an A100?
```
test_dense = tf.keras.layers.Dense(10, dtype="mixed_bfloat16")
test_dense(np.random.randn(16, 10))
```

Gives me a tensor of dtype bfloat16. Definitely would like to understand TF's behavior on this. Thanks!

**Comment by reedwm** [comment]
The MatMul in the Dense layer silently runs on the CPU, since bfloat16 isn't supported on GPUs (except in XLA). You can verify this by calling `tf.debugging.set_log_device_placement(True)` at the start of the program, which prints the device for each op run.

Note the [`tf.config.set_soft_device_placement`](https://www.tensorflow.org/api_docs/python/tf/config/set_soft_device_placement) API allows you raise an error when there is no GPU implementation instead of silently falling back to the CPU.

**Comment by cchan-lm** [comment]
Thank you @reedwm ! I now see that the CPU is used for the MatMul op, and that XLA results in placing the op on GPU as `__inference_<fn name>_<#>`.  Will try to use XLA for our purposes. For now, I consider the XLA solution as what we need, As for the original issue statement with BatchNorm's empty output, not sure if you want to keep the issue open as you investigate, but I'm okay with this issue being closed for now if you are.

**Comment by reedwm** [comment]
@cchan-lm, glad to hear XLA is a suitable workaround.

@penpornk, it turns out this is an MKL issue in [mkl_tmp_bf16_ops.cc](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/mkl/mkl_tmp_bf16_ops.cc). When FusedBatchNormV3 is run with bfloat16 and MKL is not used, a NoOp kernel is used, and so the output is empty. Instead we should raise an error. Any ideas how to fix this? Perhaps instead of using a NoOp kernel, we can use a kernel that raises an error.
